certora/specs/GuardianStorage.spec

```diff
@@ -0,0 +1,372 @@
+ * This file is derived from OwnerReach.spec for a Safe account.
+ * https://github.com/safe-global/safe-smart-account/blob/main/certora/specs/OwnerReach.spec
+ *
+ * This file uses a reach predicate:
+ *    ghost reach(address, address, address) returns bool
+ * to represent the transitive of the next
+ * relation given by the "owners" field relation for a specific wallet.
+ *
+ * The idea comes from the paper
+ *
+ * [1] Itzhaky, S., Banerjee, A., Immerman, N., Nanevski, A., Sagiv, M. (2013).
+ *     Effectively-Propositional Reasoning about Reachability in Linked Data Structures.
+ *     In: CAV 2013. Springer, https://doi.org/10.1007/978-3-642-39799-8_53
+ */
+
+using SafeHarness as safeContract;
+using SocialRecoveryModule as socialRecoveryModule;
+
+methods {
+    // Safe Functions
+    function safeContract.isModuleEnabled(address) external returns (bool) envfree;
+
+    // GuardianManager Functions
+    function threshold(address) external returns (uint256) envfree;
+    function isGuardian(address, address) external returns (bool) envfree;
+
+}
+
+persistent ghost reach(address, address, address) returns bool {
+    init_state axiom forall address X. forall address Y. forall address wallet. reach(wallet, X, Y) == (X == Y || to_mathint(Y) == 0);
+}
+
+persistent ghost mapping(address => mapping(address => address)) ghostGuardians {
+    init_state axiom forall address wallet. forall address X. to_mathint(ghostGuardians[wallet][X]) == 0;
+}
+
+persistent ghost ghostSuccCount(address, address) returns mathint {
+    init_state axiom forall address wallet. forall address X. ghostSuccCount(wallet, X) == 0;
+}
+
+persistent ghost mapping(address => uint256) ghostGuardianCount {
+    init_state axiom forall address X. to_mathint(ghostGuardianCount[X]) == 0;
+}
+
+persistent ghost address SENTINEL {
+    axiom to_mathint(SENTINEL) == 1;
+}
+
+persistent ghost address NULL {
+    axiom to_mathint(NULL) == 0;
+}
+
+// threshold can be 0 in case no Guardian is set for a wallet.
+invariant thresholdSet(address wallet)
+    threshold(wallet) <= ghostGuardianCount[wallet]
+    {
+        preserved {
+            requireInvariant reachNull();
+            requireInvariant reachInvariant();
+            requireInvariant inListReachable();
+            requireInvariant reachableInList();
+        }
+    }
+
+// every element with 0 in the guardians field can only reach the null pointer and itself
+invariant nextNull()
+    (forall address wallet. ghostGuardians[wallet][NULL] == 0) &&
+    (forall address wallet. forall address X. forall address Y. ghostGuardians[wallet][X] == 0 && reach(wallet, X, Y) => X == Y || Y == 0)
+    {
+        preserved {
+            requireInvariant reachInvariant();
+            requireInvariant inListReachable();
+            requireInvariant reachableInList();
+            requireInvariant reachNull();
+        }
+    }
+
+// every element reaches the 0 pointer (because we replace in reach the end sentinel with null)
```

Comment:
The purpose of this invariant isn't clear to me. Also, I thought that `SENTINEL` linked list was cyclic (so an address in the canonical linked list doesn't reach NULL).

Reply:
During verification, the list starts with `SENTINEL` and ends with a `NULL`. This is done by updating the ghosts in store hook. This was recommended by Certora team.